File: go/api/Voice.go
package api

import (
	"github.com/lvcdy/huawei-lte-api-go/session"
)

// Voice 对应 Voice.py。
type Voice struct {
	*session.ApiGroup
}

// NewVoice 创建 Voice API 分组。
func NewVoice(s *session.Session) *Voice {
	return &Voice{ApiGroup: session.NewApiGroup(s)}
}

// Featureswitch 对应 voice/featureswitch。
func (v *Voice) Featureswitch() (map[string]interface{}, error) {
	return v.S.Get("voice/featureswitch", nil, "api")
}

// Sipaccount 对应 voice/sipaccount。
func (v *Voice) Sipaccount() (map[string]interface{}, error) {
	return v.S.Get("voice/sipaccount", nil, "api")
}

// Sipadvance 对应 voice/sipadvance。
func (v *Voice) Sipadvance() (map[string]interface{}, error) {
	return v.S.Get("voice/sipadvance", nil, "api")
}

// Sipserver 对应 voice/sipserver。
func (v *Voice) Sipserver() (map[string]interface{}, error) {
	return v.S.Get("voice/sipserver", nil, "api")
}

// Speeddial 对应 voice/speeddial。
func (v *Voice) Speeddial() (map[string]interface{}, error) {
	return v.S.Get("voice/speeddial", nil, "api")
}

// Functioncode 对应 voice/functioncode。
func (v *Voice) Functioncode() (map[string]interface{}, error) {
	return v.S.Get("voice/functioncode", nil, "api")
}

// Voiceadvance 对应 voice/voiceadvance。
func (v *Voice) Voiceadvance() (map[string]interface{}, error) {
	return v.S.Get("voice/voiceadvance", nil, "api")
}

// Voicebusy 对应 voice/voicebusy。
func (v *Voice) Voicebusy() (map[string]interface{}, error) {
	return v.S.Get("voice/voicebusy", nil, "api")
}

// Codec 对应 voice/codec。
func (v *Voice) Codec() (map[string]interface{}, error) {
	return v.S.Get("voice/codec", nil, "api")
}

// Voiperstatus 对应 voice/voiperstatus。
func (v *Voice) Voiperstatus() (map[string]interface{}, error) {
	return v.S.Get("voice/voiperstatus", nil, "api")
}

// Volte 对应 voice/volte。
func (v *Voice) Volte() (map[string]interface{}, error) {
	return v.S.Get("voice/volte", nil, "api")
}

File: src/api/voice.rs
//! Voice API group (`api/Voice.py`).
//!
//! VoIP accounts, codecs and call features.

use serde_json::Value;

use crate::connection::Connection;
use crate::errors::Result;

use super::get_value;

/// Voice API group.
pub struct Voice<'a> {
    conn: &'a Connection,
}

impl<'a> Voice<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        Voice { conn }
    }

    /// `voice/featureswitch`. Voice feature switch.
    pub fn featureswitch(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/featureswitch")
    }

    /// `voice/sipaccount`. SIP account settings.
    pub fn sipaccount(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/sipaccount")
    }

    /// `voice/sipadvance`. Advanced SIP settings.
    pub fn sipadvance(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/sipadvance")
    }

    /// `voice/sipserver`. SIP server settings.
    pub fn sipserver(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/sipserver")
    }

    /// `voice/speeddial`. Speed-dial settings.
    pub fn speeddial(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/speeddial")
    }

    /// `voice/functioncode`. Function code settings.
    pub fn functioncode(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/functioncode")
    }

    /// `voice/voiceadvance`. Advanced voice settings.
    pub fn voiceadvance(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/voiceadvance")
    }

    /// `voice/voicebusy`. Busy handling settings.
    pub fn voicebusy(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/voicebusy")
    }

    /// `voice/codec` (reverse engineered, likely unused).
    pub fn codec(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/codec")
    }

    /// `voice/voiperstatus`. VoIP error status.
    pub fn voiperstatus(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/voiperstatus")
    }

    /// `voice/volte`. VoLTE state.
    pub fn volte(&self) -> Result<Value> {
        get_value(self.conn, "api/voice/volte")
    }
}
